builder/proxmox/common/step_download_iso_on_pve.go

```diff
@@ -0,0 +1,97 @@
+type stepDownloadISOOnPVE struct {
+	ISO *additionalISOsConfig
+}
+
+func (s *stepDownloadISOOnPVE) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	var isoStoragePath string
+	isoStoragePath = Download_iso_on_pve(state, s.ISO.ISOUrls, s.ISO.ISOChecksum, s.ISO.ISOStoragePool)
+
+	// If available, set the file path to the downloaded iso file on the node
+	if isoStoragePath != "" {
+		s.ISO.ISOFile = isoStoragePath
+		return multistep.ActionContinue
+	}
+	// Abort if no ISO can be downloaded
+	return multistep.ActionHalt
+}
+
+// Download_iso_on_pve abstracts the checksum and download process os that the code can be shared between
+// the common module and the iso module. This is necessary because both handle the storage path to the iso differently.
+//
+// The function takes a list of URLs to download the iso and tries them one another.
+// If a download was successful it skips the additonal downlaod mirrors and returns the path to the iso on the node.
+//
+// Returns: When successful, the path to the iso on the node, else an empty string.
+func Download_iso_on_pve(state multistep.StateBag, ISOUrls []string, ISOChecksum string, ISOStoragePool string) string {
+	ui := state.Get("ui").(packersdk.Ui)
+	client := state.Get("proxmoxClient").(*proxmox.Client)
+	c := state.Get("config").(*Config)
+
+	// Generate ISOConfig configuration attributes in the format defined for packer-plugin-sdk
+	// and use go-getter to generate parameters compatible with the Proxmox-API.
+	var isoConfig proxmox.ConfigContent_Iso
+	for _, url := range ISOUrls {
+		var checksum string
+		var checksumType string
+		if ISOChecksum != "none" {
+			gr := &getter.Request{
+				Src: url + "?checksum=" + ISOChecksum,
+			}
+			gc := getter.Client{}
+			fileChecksum, err := gc.GetChecksum(context.TODO(), gr)
+			if err != nil {
+				state.Put("error", err)
+				ui.Error(err.Error())
+				continue
+			}
+			checksum = hex.EncodeToString(fileChecksum.Value)
+			checksumType = fileChecksum.Type
+		}
+
+		isoConfig = proxmox.ConfigContent_Iso{
+			Node:              c.Node,
+			Storage:           ISOStoragePool,
+			DownloadUrl:       url,
+			Filename:          path.Base(url),
+			ChecksumAlgorithm: checksumType,
+			Checksum:          checksum,
+		}
+
+		ui.Say(fmt.Sprintf("Beginning download of %s to node %s", isoConfig.DownloadUrl, isoConfig.Node))
+		err := proxmox.DownloadIsoFromUrl(client, isoConfig)
+		// On error continues with the next URL and logs the error
+		if err != nil {
+			state.Put("error", err)
+			ui.Error(err.Error())
+			continue
+		}
+		isoStoragePath := fmt.Sprintf("%s:iso/%s", isoConfig.Storage, isoConfig.Filename)
+		ui.Say(fmt.Sprintf("Finished downloading %s", isoStoragePath))
+		// Returns the path to the iso on the node
+		return isoStoragePath
+	}
+	// Returns an empty string, which means download was not successful.
+	return ""
+}
+
+func (s *stepDownloadISOOnPVE) Cleanup(state multistep.StateBag) {
```

Comment:
I wonder, should we clean the downloaded ISOs from the server? Is there an API call to do this?

Reply:
I can look into it, but I would suggest keeping it behind a flag that enables or disables the cleanup. I don't think we have to delete the e. g. virtio drivers every time as it can be reused in the next template. At least an my use case of installing multiple windows versions it is faster when keeping the files.

Reply:
It seems to be possible but I think that would be a new feature as it is also not done for the normal upload of iso files. The only iso file that gets deleted is the auto-generated one in `stepUploadAdditionalISO`. So I think this could be one solution for all these procedures and not specific to this step.